File: src/solar/solar-common/util/frame/IconChooserUtil.ts
import {Frame} from "@/frame";
import FramePoint from "@/FramePoint";
import TextAlign from "@/TextAlign";
import AsyncUtil from "@/AsyncUtil";

declare global {

    interface IconOption {
        /**图标项目名字*/
        name: string
        /**图标项目图标*/
        icon?: string,
        /**图标项目信息*/
        info?: string,
    }

}

/**
 * 图标选择工具(比如三选一)
 * 可通过此工具解耦显示与需要显示的功能耦合
 * 切换显示效果可以重新设置IconChooserUtil.show = 你自己的显示函数
 */
export default class IconChooserUtil {
    static iconWidth = 0.12
    static iconHeight = 0.24
    static iconGap = 0.04
    static centerY = 0.35


    /**
     * 需要再同步逻辑中调用
     * @param playerId 给指定玩家显示选项
     * @param iconOptions 显示选项的备选项
     * @param onClickCallBack 当点击后的回调 (同步的)
     * @param title 可选的标题文字
     */
    static show(playerId: number, iconOptions: IconOption[], onClickCallBack: (index: number, iconOption: IconOption) => void, title?: string): void {

        //异步执行
        let startX = (0.8 - (IconChooserUtil.iconWidth * iconOptions.length + IconChooserUtil.iconGap * (iconOptions.length - 1))) / 2
        let rootFrame = new Frame("FRAME");
        for (let i = 0; i < iconOptions.length; i++) {
            let iconOption = iconOptions[i];
            let backDropFrame = Frame.createBackDrop(rootFrame.current);
            backDropFrame.setTexture(iconOption.icon)
            backDropFrame.setSize(IconChooserUtil.iconWidth, IconChooserUtil.iconHeight);
            backDropFrame.setAbsPoint(FramePoint.left, startX + ((IconChooserUtil.iconWidth + 0.04) * i), IconChooserUtil.centerY);


            let nameFrame = Frame.createTEXT(backDropFrame.current);
            nameFrame.setText(iconOption.name)
            nameFrame.setPoint(FramePoint.top, backDropFrame.current, FramePoint.top, 0, -0.01)
            nameFrame.setTextAlignment(TextAlign.top)

            let infoFrame = Frame.createTEXT(backDropFrame.current);
            infoFrame.setText(iconOption.info)
            infoFrame.setAllPoints(backDropFrame.current)
            infoFrame.setTextAlignment(TextAlign.center)

            let btn = Frame.createGLUEBUTTON(backDropFrame.current)
            btn.setAllPoints(backDropFrame.current)
            let index = i;
            btn.setOnClick(() => {
                rootFrame.setVisible(false);
                DzDestroyFrame(rootFrame.current);
                onClickCallBack(index, iconOption);
            }, true);

            backDropFrame.setVisible(false)
            AsyncUtil.run(() => {
                backDropFrame.setVisible(true)
            }, Player(playerId));
        }

        if (title) {
            AsyncUtil.run(() => {
                let titleFrame = Frame.createTEXT(rootFrame.current);
                titleFrame.setWidth(0.3)
                titleFrame.setText(title)
                titleFrame.setFont(0.026)
                titleFrame.setAbsPoint(FramePoint.top, 0.4, IconChooserUtil.centerY + (IconChooserUtil.iconHeight / 2) + IconChooserUtil.iconGap)
                titleFrame.setTextAlignment(TextAlign.top)
            }, Player(playerId));
        }
    }


}

File: scripts/solar/solar-common/util/frame/IconChooserUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 1,["8"] = 1,["9"] = 2,["10"] = 2,["11"] = 3,["12"] = 3,["13"] = 4,["14"] = 4,["15"] = 24,["16"] = 24,["17"] = 24,["19"] = 24,["20"] = 38,["21"] = 41,["22"] = 42,["24"] = 43,["25"] = 43,["26"] = 44,["27"] = 45,["28"] = 46,["29"] = 47,["30"] = 48,["31"] = 51,["32"] = 52,["33"] = 53,["34"] = 53,["35"] = 53,["36"] = 53,["37"] = 53,["38"] = 53,["39"] = 53,["40"] = 54,["41"] = 56,["42"] = 57,["43"] = 58,["44"] = 59,["45"] = 61,["46"] = 62,["47"] = 63,["48"] = 64,["49"] = 64,["50"] = 65,["51"] = 66,["52"] = 67,["53"] = 64,["54"] = 64,["55"] = 64,["56"] = 70,["57"] = 71,["58"] = 71,["59"] = 72,["60"] = 71,["61"] = 71,["62"] = 71,["63"] = 43,["66"] = 76,["67"] = 77,["68"] = 77,["69"] = 78,["70"] = 79,["71"] = 80,["72"] = 81,["73"] = 82,["74"] = 83,["75"] = 77,["76"] = 77,["77"] = 77,["79"] = 38,["80"] = 25,["81"] = 26,["82"] = 27,["83"] = 28});
local ____exports = {}
local ____frame = require("solar.solar-common.w3ts.handles.frame")
local Frame = ____frame.Frame
local ____FramePoint = require("solar.solar-common.constant.FramePoint")
local FramePoint = ____FramePoint.default
local ____TextAlign = require("solar.solar-common.constant.TextAlign")
local TextAlign = ____TextAlign.default
local ____AsyncUtil = require("solar.solar-common.util.net.AsyncUtil")
local AsyncUtil = ____AsyncUtil.default
____exports.default = __TS__Class()
local IconChooserUtil = ____exports.default
IconChooserUtil.name = "IconChooserUtil"
function IconChooserUtil.prototype.____constructor(self)
end
function IconChooserUtil.show(self, playerId, iconOptions, onClickCallBack, title)
    local startX = (0.8 - (____exports.default.iconWidth * #iconOptions + ____exports.default.iconGap * (#iconOptions - 1))) / 2
    local rootFrame = __TS__New(Frame, "FRAME")
    do
        local i = 0
        while i < #iconOptions do
            local iconOption = iconOptions[i + 1]
            local backDropFrame = Frame:createBackDrop(rootFrame.current)
            backDropFrame:setTexture(iconOption.icon)
            backDropFrame:setSize(____exports.default.iconWidth, ____exports.default.iconHeight)
            backDropFrame:setAbsPoint(FramePoint.left, startX + (____exports.default.iconWidth + 0.04) * i, ____exports.default.centerY)
            local nameFrame = Frame:createTEXT(backDropFrame.current)
            nameFrame:setText(iconOption.name)
            nameFrame:setPoint(
                FramePoint.top,
                backDropFrame.current,
                FramePoint.top,
                0,
                -0.01
            )
            nameFrame:setTextAlignment(TextAlign.top)
            local infoFrame = Frame:createTEXT(backDropFrame.current)
            infoFrame:setText(iconOption.info)
            infoFrame:setAllPoints(backDropFrame.current)
            infoFrame:setTextAlignment(TextAlign.center)
            local btn = Frame:createGLUEBUTTON(backDropFrame.current)
            btn:setAllPoints(backDropFrame.current)
            local index = i
            btn:setOnClick(
                function()
                    rootFrame:setVisible(false)
                    DzDestroyFrame(rootFrame.current)
                    onClickCallBack(nil, index, iconOption)
                end,
                true
            )
            backDropFrame:setVisible(false)
            AsyncUtil:run(
                function()
                    backDropFrame:setVisible(true)
                end,
                Player(playerId)
            )
            i = i + 1
        end
    end
    if title then
        AsyncUtil:run(
            function()
                local titleFrame = Frame:createTEXT(rootFrame.current)
                titleFrame:setWidth(0.3)
                titleFrame:setText(title)
                titleFrame:setFont(0.026)
                titleFrame:setAbsPoint(FramePoint.top, 0.4, ____exports.default.centerY + ____exports.default.iconHeight / 2 + ____exports.default.iconGap)
                titleFrame:setTextAlignment(TextAlign.top)
            end,
            Player(playerId)
        )
    end
end
IconChooserUtil.iconWidth = 0.12
IconChooserUtil.iconHeight = 0.24
IconChooserUtil.iconGap = 0.04
IconChooserUtil.centerY = 0.35
return ____exports
